opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp

```diff
@@ -109,9 +109,10 @@ namespace Opm
         const Events& getEvents() const;
         const Deck& getModifierDeck(size_t timeStep) const;
         bool hasOilVaporizationProperties() const;
-        const std::map<int, VFPProdTable>& getVFPProdTables(size_t timeStep) const;
-        const std::map<int, VFPInjTable>& getVFPInjTables(size_t timeStep) const;
-
+        const VFPProdTable& getVFPProdTable(int table_id, size_t timeStep) const;
+        const VFPInjTable& getVFPInjTable(int table_id, size_t timeStep) const;
+        std::map<int, VFPProdTable> getVFPProdTables(size_t timeStep) const;
```

Comment:
@joakim-hove, I guess you probably will not like `std::map<int, const VFPProdTable*>` or something similar here.  But since `VFPProdTable` is a temporary variable here, with the current master branch, the information in the table is lost after the `VFPProperties` is created. The easiest way to me is to use some kind of Pointer to refer to the `VFPProdTable` in the `DynamicState`. And also, all the `VFPProperties` related was implemented with Pointers. Basically there will be some fighting if we use `std::map<int, VFPProdTable>` here.  Let me know if it is okay to you to change here to be a map of `Pointer`.

Reply:
I am not scared of pointers as such; but I do not fully understand what you want to achieve here:

**Lifetime:** Does the `VFProperties` object need the `VFPProdTable` to stay alive? To me the impression was that the `VFPProdTable` could safely go out of scope after the `VFPProperties` was constructed (that is the way it *should* be in my opinion) - but these lifetime expectations were not really clear to me - I might have got it wrong?

**Schedule Events:** My assumption is that the VFP tables changes seldom, typically I guess only once in the transition from history to prediction. My intention was therefore that simulator would check the `ScheduleEvents::VFPINJ_UPDATE` and `ScheduleEvents::VFPPROD_UPDATE` events and then do a full VFP reinitialization when the events are fired.

Anyway - the current API in opm-common is of course not written in stone, it was a first attempt which I thought would be convenient - if *the* consumer thinks otherwise we change it!

Reply:
> Does the `VFProperties` object need the `VFPProdTable` to stay alive? 

Yes. In the implementation of `VFProperties` (including `VFPProdProperties` and `VFPInjProperties`, unless we want to put effort to refactor it, while it is not why I am checking VFP related. The reason is that it does not work for simulation), it points to the `VFPProdTable` or `VFPInjTable` in the `EclipseState` (in `TableManager` before we changed it.), probably also to avoid copying them.  To continue using the same `VFPproperties`, we need to find a way to let it point to the tables in the `Schedule`.  But in https://github.com/OPM/opm-simulators/pull/1450 , I changed the code to let `VFPProperties` related to point to the temporary `VFPProdTable`s and `VFPInjTable`s in the `std::map`, which I was thinking just an interface change. I was wrong.

One way to be over this is to make the `std::map` a copy not an map of Pointers towards VFP tables in `EclipseState` anymore, while it is basically a refactoring of the class `VFPProperties`, and I am not sure what gonna come up with all the fighting came up last time.  The other easier way is to return `std::map<int, VFPProdTable*>`, and `VFPProdTable *` points to the corresponding tables in `Schedule`. 

 > but to me it seems a bit like you are trying to emulate DynamicState<> also for the VFPProperties?

Although things changed some due to different opinions, the things (most) in `BlackoilWellModel` (owner of `VFPProperties`)  are supposed to only survive for one report step (in theory). I am still learning `VFP` related implementation. If you think the implementation not good (I might have implemented it differently too), we can discuss the refactoring in later PRs, maybe both on the parser and simulator side, I would be willing to join. But my current task is to fix the running, the easiest way is to use `std::map<int, VFPProdTable*>` here, which includes a few lines of change in `opm-common`.